### PR TITLE
[Model] Whisper model implementation

### LINKED ISSUE #180: Whisper support
Is support for Whisper on the roadmap? Something like https://github.com/ggerganov/whisper.cpp would be great.

### PR DESCRIPTION
**Pull request by aurickq**
Add Whisper model implementation. Based on https://github.com/vllm-project/vllm/pull/5964 but heavily optimized and integrated with newer encoder/decoder support in vLLM. Currently only supports audio up to 30s. No VAD model, no beam search, no phoneme model, forced alignment, etc.

Performance overall looks OK, especially at larger batch sizes

Screenshot 2024-12-18 at 5 33 15 PM

FIX https://github.com/vllm-project/vllm/issues/180

### CODE REVIEW

File: vllm/worker/enc_dec_model_runner.py [hunk 1]
```diff
@@ -280,19 +280,19 @@ def profile_run(self) -> None:
         for group_id in range(max_num_seqs):
             seq_len = (max_num_batched_tokens // max_num_seqs +
                        (group_id < max_num_batched_tokens % max_num_seqs))
+            seq_len = min(seq_len, self.model_config.max_model_len)
```
Comment (DarkLight1337) [inline vllm/worker/enc_dec_model_runner.py:283]: Hmm, I think `seq_len` should never be calculated to be greater than `max_model_len` in the first place.

File: vllm/transformers_utils/tokenizer_group/tokenizer_group.py [hunk 2]
```diff
@@ -57,7 +59,11 @@ def encode(self,
                request_id: Optional[str] = None,
                lora_request: Optional[LoRARequest] = None) -> List[int]:
```
Comment (DarkLight1337) [inline vllm/transformers_utils/tokenizer_group/tokenizer_group.py:60]: To be consistent with HF, I think `add_special_tokens` should be an argument to `encode` and `encoder_async` instead of the initializer.

File: vllm/model_executor/models/registry.py [hunk 3]
```diff
@@ -165,6 +165,7 @@
     "UltravoxModel": ("ultravox", "UltravoxModel"),
     # [Encoder-decoder]
     "MllamaForConditionalGeneration": ("mllama", "MllamaForConditionalGeneration"),  # noqa: E501
+    "WhisperForConditionalGeneration": ("whisper", "WhisperForConditionalGeneration"),  # noqa: E501
```
Comment (DarkLight1337) [inline vllm/model_executor/models/registry.py:168]: Please update the Supported Models page and `tests/models/registry.py` accordingly.

File: vllm/inputs/preprocess.py [hunk 4]
```diff
@@ -439,8 +439,15 @@ def _build_enc_dec_llm_inputs(
...
-                None)
+            if self.model_config.hf_config.model_type == "whisper":
+                # For Whisper models, the text prompt should go to the decoder.
+                # If no explicit encoder/decoder inputs, then copy the prompt
+                # from the encoder to the decoder. The encoder tokens are later
+                # overridden by the audio features.
+                dec_token_ids = encoder_inputs["prompt_token_ids"].copy()
+            else:
+                dec_token_ids = self._prepare_decoder_input_ids_for_generation(
+                    None)
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py:450]: Is there a way to determine this without model type information?

File: vllm/model_executor/models/whisper.py [hunk 5]
```diff
@@ -0,0 +1,689 @@
...
+                       padding=False, return_tensors="pt")
+    kwargs["input_features"] = kwargs["input_features"].squeeze(0)
+    kwargs["input_features"] = kwargs["input_features"].to(torch.float16)
+
+    return MultiModalKwargs(kwargs)
+
+
+@INPUT_REGISTRY.register_dummy_encoder_data(dummy_encoder_data_for_whisper)
+@INPUT_REGISTRY.register_input_processor(input_processor_for_whisper)
+@MULTIMODAL_REGISTRY.register_input_mapper("audio", input_mapper_for_whisper)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/whisper.py:625]: If possible, use the merged multi-modal processor (#10114) which should simplify your code.

**Review by DarkLight1337 — COMMENTED** [review]
Some initial comments.

File: vllm/model_executor/models/whisper.py [hunk 6]
```diff
@@ -0,0 +1,689 @@
...
+    if channels % 2 != 0:
+        raise ValueError(
+            f"Number of channels has to be divisible by 2 for sinusoidal "
+            f"positional embeddings, got {channels} channels."
+        )
+    log_timescale_increment = math.log(max_timescale) / (channels // 2 - 1)
+    inv_timescales = torch.exp(-log_timescale_increment *
+                               torch.arange(channels // 2))
+    scaled_time = torch.arange(length).view(-1, 1) * inv_timescales.view(1, -1)
+    return torch.cat([scaled_time.sin(), scaled_time.cos()], dim=1)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/whisper.py:53]: Let's import from `transformers` if the implementation is exactly the same as theirs.

File: vllm/model_executor/models/whisper.py [hunk 7]
```diff
@@ -0,0 +1,689 @@
...
+        self.embed_dim = config.d_model
+        self.self_attn = WhisperAttention(
+            embed_dim=self.embed_dim,
+            num_heads=config.decoder_attention_heads,
+            attn_type=AttentionType.DECODER,
+            cache_config=cache_config,
+            quant_config=quant_config,
+            prefix=f"{prefix}.self_attn",
+        )
+        self.activation_fn = FastGELU()
```
Comment (DarkLight1337) [inline vllm/model_executor/models/whisper.py:303]: Do we not support other activation functions?

File: vllm/model_executor/models/whisper.py [hunk 8]
```diff
@@ -0,0 +1,689 @@
...
+                                       config.d_model,
+                                       quant_config=quant_config)
+        self.proj_out = self.proj_out.tie_weights(
+            self.model.decoder.embed_tokens)
+        logit_scale = getattr(config, "logit_scale", 1.0)
+        self.logits_processor = LogitsProcessor(self.unpadded_vocab_size,
+                                                config.vocab_size, logit_scale)
+        self.sampler = Sampler()
+
+    def forward(
```
Comment (DarkLight1337) [inline vllm/model_executor/models/whisper.py:650]: To support V1, the model needs to implement `get_input_embeddings` and `get_multimodal_embeddings`.

**Review by Isotr0py — COMMENTED** [review]
Just some initial comments about model implementation :)

File: vllm/model_executor/models/whisper.py (hunk 7 above)
Comment (aurickq) [reply vllm/model_executor/models/whisper.py]: use `config.activation_function` now

File: requirements-common.txt [hunk 9]
```diff
@@ -35,3 +35,4 @@ setuptools>=74.1.1; python_version > '3.11' # Setuptools is used by triton, we n
 einops # Required for Qwen2-VL.
 compressed-tensors == 0.8.0 # required for compressed-tensors
 depyf==0.18.0 # required for profiling and debugging torch.compile
+librosa >= 0.10.2  # required for audio processing including Whisper
```
Comment (mgoin) [inline requirements-common.txt:38]: I believe we made librosa optional in the past due to licensing concerns, I can find the issue that started this if needed

We currently have it as a dep through the vllm[audio] extra here https://github.com/vllm-project/vllm/blob/fdea8ec16775e1645620b5ff46b799d60df4624c/setup.py#L633

File: vllm/model_executor/models/whisper.py [hunk 10]
```diff
@@ -0,0 +1,702 @@
+import math
+from functools import lru_cache
+from typing import Iterable, List, Mapping, Optional, Set, Tuple, Union
+
+import librosa
```
Comment (mgoin) [inline vllm/model_executor/models/whisper.py:5]: Related to librosa in the audio extra, we should make this a lazy import

File: vllm/transformers_utils/tokenizer_group/tokenizer_group.py [hunk 11]
```diff
@@ -55,19 +55,31 @@ def _raise_if_input_too_long(self,
...
                request_id: Optional[str] = None,
-               lora_request: Optional[LoRARequest] = None) -> List[int]:
+               lora_request: Optional[LoRARequest] = None,
+               add_special_tokens: Optional[bool] = None) -> List[int]:
         tokenizer = self.get_lora_tokenizer(lora_request)
-        ret = tokenizer.encode(prompt)
+        if add_special_tokens is not None:
+            ret = tokenizer.encode(  # type: ignore[call-arg]
+                prompt,
+                add_special_tokens=add_special_tokens)
```
Comment (DarkLight1337) [inline vllm/transformers_utils/tokenizer_group/tokenizer_group.py:64]: Let's move `vllm.multmodal.processor._encode` to `vllm.transformers_utils` so we can reuse the logic and avoid this type error.

File: vllm/model_executor/models/whisper.py (hunk 5 above)
Comment (aurickq) [reply vllm/model_executor/models/whisper.py:646]: What's the right approach here? I tried the Processor and there appears to be several mismatched assumptions with Whisper and encoder-decoder models:

1. There is a mandatory `_get_prompt_replacements` which does not apply to Whisper (audio tokens never get embedded within text tokens).
2. The processor assumes one token sequence but for Whisper, two token sequences are needed: one as a placeholder for the encoder to allocate the correct cross-attention kv-cache, and the second for the actual prompt which only goes to the decoder.

(1) is easy to get around with some awkwardness, but (2) seems only supported by the legacy multimodal processing pipeline which returns two different inputs, one for the encoder and another for the decoder.

File: vllm/inputs/preprocess.py (hunk 4 above)
Comment (aurickq) [reply vllm/inputs/preprocess.py:463]: I am not sure about generalizing this from a single example. In the long term it may be better to allow the model definition to specify exactly the mapping between input fields and where they go (e.g. encoder/decoder)

File: vllm/model_executor/models/whisper.py (hunk 5 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/whisper.py:646]: Thanks for bringing this up! I haven't thought about encoder-decoder models for the new pipeline yet... let's use the legacy processing pipeline for now then.

File: vllm/model_executor/models/whisper.py [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+    )
+
+
+def _resample(audio: np.ndarray, orig_sr: int, target_sr: int) -> np.ndarray:
+    try:
+        import librosa
+    except ImportError as exc:
+        raise ImportError(
+            "Please install vllm[audio] for audio support.") from exc
+    return librosa.resample(audio, orig_sr=orig_sr, target_sr=target_sr)
```
Comment (Isotr0py) [inline vllm/model_executor/models/whisper.py:591]: Seems that this is totally same with the `resample_audio` in `vllm/multimodal/audio.py`?

File: vllm/model_executor/models/whisper.py [hunk 13]
```diff
@@ -0,0 +1,711 @@
...
+    **kwargs,
+) -> WhisperProcessor:
+    """Gets an whisper processor for the given model name via HuggingFace."""
+    return WhisperProcessor.from_pretrained(
+        processor_name,
+        *args,
+        trust_remote_code=trust_remote_code,
+        revision=revision,
+        **kwargs,
+    )
```
Comment (Isotr0py) [inline vllm/model_executor/models/whisper.py:582]: Perhaps we can use `cached_get_processor` from `vllm/transformers_utils/processor.py`?

File: vllm/model_executor/models/whisper.py (hunk 8 above)
Comment (aurickq) [reply vllm/model_executor/models/whisper.py:666]: Done, but there is some awkwardness
- `get_multimodal_embeddings` for Whisper also requires kv-cache and attn metadata which is not in the base method's interface
- It looks like `get_input_embeddings` expects to embed multimodal tokens into text tokens. again, this is not what Whisper does. Currently I just have this method return the decoder token embeddings.

I couldn't find anywhere in vLLM that uses these two methods so it's hard for me to understand their intended usage. Happy to change the implementation to whatever works better.

Comment (DarkLight1337) [reply vllm/model_executor/models/whisper.py:666]: Sorry, I forgot that V1 doesn't support encoder decoder yet. Let's just add a TODO and come back to this later.

Comment (aurickq) [reply vllm/model_executor/models/whisper.py:666]: added todos

**Comment by orgh0** [comment]
Hey @sfc-gh-aqiao  , thanks so much for taking this forward. It's going to help a lot of folks. 

Quick question:- By when do you estimate this feature would get merged? 
trying to understand if i could start using this branch for the work we do. Given it's close to merging.

**Comment by aurickq** [comment]
> Hey @sfc-gh-aqiao , thanks so much for taking this forward. It's going to help a lot of folks.
> 
> Quick question:- By when do you estimate this feature would get merged? trying to understand if i could start using this branch for the work we do. Given it's close to merging.

Just fixed a few minor issues and rebased. It should be ready to go, and it works well in my tests. Just waiting on approval or further reviews.

File: vllm/model_executor/models/whisper.py [hunk 14]
```diff
@@ -0,0 +1,736 @@
...
+
+
+def dummy_encoder_data_for_whisper(ctx: InputContext, seq_len: int,
+                                   mm_counts: Mapping[str, int]):
+    assert mm_counts["audio"] == 1
+    sample_rate = 16000
+    num_tokens = ctx.model_config.hf_config.max_source_positions
+    return DummyData(
+        SequenceData.from_prompt_token_counts((0, num_tokens)),
+        {"audio": [(np.zeros(30 * sample_rate), sample_rate)]},
```
Comment (mgoin) [inline vllm/model_executor/models/whisper.py:587]: Is the 30 seconds here just hardcoded in the model, or is there some place we can read for this limit?

File: vllm/model_executor/models/whisper.py [hunk 15]
```diff
@@ -0,0 +1,736 @@
...
+
+def get_max_whisper_audio_tokens(ctx: InputContext) -> int:
+    return ctx.model_config.hf_config.max_source_positions
+
+
+def dummy_encoder_data_for_whisper(ctx: InputContext, seq_len: int,
+                                   mm_counts: Mapping[str, int]):
+    assert mm_counts["audio"] == 1
+    sample_rate = 16000
+    num_tokens = ctx.model_config.hf_config.max_source_positions
```
Comment (mgoin) [inline vllm/model_executor/models/whisper.py:584]: nit: 
```suggestion
    num_tokens = get_max_whisper_audio_tokens(ctx)
```

File: vllm/model_executor/models/whisper.py [hunk 16]
```diff
@@ -0,0 +1,736 @@
...
+        assert len(multi_modal_data["audio"]) == 1
+        multi_modal_data["audio"] = multi_modal_data["audio"][0]
+    # Resample and process audio
+    audio, orig_sr = multi_modal_data["audio"]
+    processor = cached_get_processor(ctx.model_config.model)
+    target_sr = processor.feature_extractor.sampling_rate
+    audio = resample_audio(audio, orig_sr=orig_sr, target_sr=target_sr)
+    multi_modal_data["audio"] = (audio, target_sr)
+    # Pre-allocate placeholder tokens in encoder sequence
+    num_tokens = ctx.model_config.hf_config.max_source_positions
```
Comment (mgoin) [inline vllm/model_executor/models/whisper.py:603]:
```suggestion
    num_tokens = get_max_whisper_audio_tokens(ctx)
```

File: vllm/model_executor/models/whisper.py [hunk 17]
```diff
@@ -0,0 +1,736 @@
...
+    processor = cached_get_processor(ctx.model_config.model)
+    sampling_rate = processor.feature_extractor.sampling_rate
+
+    audios = [audio for audio, _ in multi_modal_data]
+
+    kwargs = processor(audios,
+                       sampling_rate=sampling_rate,
+                       return_tensors="pt")
+    kwargs["input_features"] = kwargs["input_features"].squeeze(0)
+    kwargs["input_features"] = kwargs["input_features"].to(torch.float16)
```
Comment (mgoin) [inline vllm/model_executor/models/whisper.py:629]: Do we support running a whisper model in bfloat16?

**Review by mgoin — COMMENTED** [review]
Thanks for the nice work! I think this should be ready to go by me - please add the test to the buildkite pipeline to verify

File: tests/models/encoder_decoder/audio/test_whisper.py
[the remark this answers is not in the document]
Comment (aurickq) [reply tests/models/encoder_decoder/audio/test_whisper.py:1]: done, thanks

File: vllm/model_executor/models/whisper.py (hunk 14 above)
Comment (aurickq) [reply vllm/model_executor/models/whisper.py]: got rid of hard coding

File: vllm/model_executor/models/whisper.py (hunk 17 above)
Comment (aurickq) [reply vllm/model_executor/models/whisper.py]: yes, got rid of this hard coding

File: vllm/inputs/preprocess.py (hunk 4 above)
Comment (afeldman-nm) [reply vllm/inputs/preprocess.py:463]: Agreed with @aurickq , long-term it is probably best to either (1) have the model definition specify whether to map the input text prompt to the encoder, or (2) add a default behavior only for *multi-modal models with cross-attention*, wherein the text prompt is always routed to the decoder & the non-text modality is always mapped to the encoder.

(I worked on adding encoder/decoder cross-attention support to v0)

File: .buildkite/test-pipeline.yaml [hunk 18]
```diff
@@ -363,12 +363,14 @@ steps:
   - tests/models/decoder_only/audio_language
   - tests/models/decoder_only/vision_language
   - tests/models/embedding/vision_language
+  - tests/models/encoder_decoder/audio
```
[the remark this answers is not in the document]
Comment (aurickq) [reply .buildkite/test-pipeline.yaml]: done

File: tests/models/encoder_decoder/audio/test_whisper.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+    for output, expected in zip(outputs, expected_list):
+        print(output.outputs[0].text)
+        assert output.outputs[0].text == expected
+
+
+@fork_new_process_for_each_test
+@pytest.mark.core_model
+@pytest.mark.parametrize("model",
+                         ["openai/whisper-medium", "openai/whisper-large-v3"])
```
[the remark this answers is not in the document]
Comment (aurickq) [reply tests/models/encoder_decoder/audio/test_whisper.py]: changed to `small` (~200M) and `large-v3-turbo` (~800M). the tiny model outputs are poor so not sure if it's worth testing. also nice to have a more recent model (`large-v3-turbo`)

**Review by mgoin — APPROVED** [review]
LGTM and verified the example locally. Last part is adding it to the model docs, feel free to do in a followup. cc @DarkLight1337 @Isotr0py for final pass

File: tests/models/registry.py [hunk 20]
```diff
@@ -128,6 +128,7 @@ class _HfExamplesInfo:
     # [Encoder-decoder]
     "BartModel": _HfExamplesInfo("facebook/bart-base"),
     "BartForConditionalGeneration": _HfExamplesInfo("facebook/bart-large-cnn"),
+    "WhisperForConditionalGeneration": _HfExamplesInfo("openai/whisper-large-v3"),  # noqa: E501
```
Comment (DarkLight1337) [inline tests/models/registry.py:131]: Let's move this under Mllama since it's a multimodal model.

File: vllm/inputs/preprocess.py [hunk 21]
```diff
@@ -184,10 +184,16 @@ def _tokenize_prompt(
         corresponding token IDs.
         """
         tokenizer = self.get_tokenizer_group()
-
+        add_special_tokens = None
+        if self.model_config.hf_config.model_type == "whisper":
+            # For Whisper, special tokens should be provided by the user based
+            # on the task and language of their request. Also needed to avoid
+            # appending an EOS token to the prompt which disrupts generation.
+            add_special_tokens = False
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py:192]: Can we remove the eos token in the input processor? But maybe we can improve this later via the merged multi-modal processor since I assume HF can handle this automatically.

Comment (Isotr0py) [reply vllm/inputs/preprocess.py:192]: Seems that `<|startoftranscript|>` is also a special token which will be added by tokenizer by default:
```python3
from transformers import AutoProcessor

processor = AutoProcessor.from_pretrained("openai/whisper-small")
tokenizer = processor.tokenizer
encode = tokenizer("<|startoftranscript|>")
print(tokenizer.decode(encode["input_ids"]))
```

It outputs `'<|startoftranscript|><|notimestamps|><|startoftranscript|><|endoftext|>'`, which means that we need to remove a `"<|startoftranscript|>"` to get the original prompt.

Perhaps we need to construct user prompts with non special tokens.

**Review by DarkLight1337 — APPROVED** [review]
Let's merge this first then.

**Comment by DarkLight1337** [comment]
Can you merge from main to fix the CI failures?

**Review by Isotr0py — APPROVED** [review]

File: vllm/inputs/preprocess.py (hunk 21 above)
Comment (aurickq) [reply vllm/inputs/preprocess.py:192]: Yes, there are other special tokens that are added as well. There are also some other challenges
1. Some tokens are not reliably mapped to the same ids. E.g. `<|notimestamps|>` is 50363 for whisper-small but 50364 for whisper-large-v3. This makes it awkward to try to detect these special tokens from inside the input processor.
2. From in the input processor, not sure how to reliably differentiate between the two cases
  a. User passed in text `prompt` which is tokenized in vLLM. We need to strip special tokens.
  b. User passed in tokens `prompt_token_ids` which should not be stripped.

Basically, I'm not sure how to reliably remove special tokens after it's already encoded. It seems error-prone and brittle to third-party behavior in huggingface. I understand that having the tokenizer not add special tokens in the first place involves more changes inside vLLM, but IMO it's the most reliable solution that's more robust to future changes in vLLM and huggingface.

**Comment by DarkLight1337** [comment]
Thanks again for your hard work!

**Comment by ArmykOliva** [comment]
Thank you for your work. Are Rest API endpoints reachable as well? Could you add an example for that please?

**Comment by aryan1165** [comment]
Just a quick question, Is anyone implementing VAD, to support chunking and accepting audios of len more than 30s. Also how continuous batching can be supported?

**Comment by Temirulan** [comment]
Thanks for great news! Few questions here:

- [ ] Is word timestamping working using this upload?
- [ ] Does this support inflight batching?

**Comment by aryan-sarvam** [comment]
Hey, I tried to use whisper in online way, using vllm.entrypoints.openai api server.py but, we can't hit Completions API (Can't send multimodal data), as mentioned in the Docs (generate and Completions API are similar), and chat_utils.py doesnt have whisper support. 

Is someone working on that part. I tried some rough solutions but couldn't get to execute properly. 

I will also open an issue regarding this, if this is not yet on the pipeline someone building.
Thanks

**Comment by aryan-sarvam** [comment]
> Thank you for your work. Are Rest API endpoints reachable as well? Could you add an example for that please?

Hi, I dont think they are, Completions one is bound to support this but it doesn't have multi modal support, and chat doesn't have this support. Let me know if you came up with a solution?

**Comment by aryan-sarvam** [comment]
> Add Whisper model implementation. Based on #5964 but heavily optimized and integrated with newer encoder/decoder support in vLLM. Currently only supports audio up to 30s. No VAD model, no beam search, no phoneme model, forced alignment, etc.
> 
> Performance overall looks OK, especially at larger batch sizes
> 
> Screenshot 2024-12-18 at 5 33 15 PM
> FIX #180

Hey, I was doing benchmarking on H100, of audio len around 18s. I found out that RTFx maxes out at a really low value around 120 at batch size 96. Can you share the benchmarking parameters or flags so that we can reproduce these results.

**Comment by aurickq** [comment]
> > Add Whisper model implementation. Based on #5964 but heavily optimized and integrated with newer encoder/decoder support in vLLM. Currently only supports audio up to 30s. No VAD model, no beam search, no phoneme model, forced alignment, etc.
> > Performance overall looks OK, especially at larger batch sizes
> > Screenshot 2024-12-18 at 5 33 15 PM
> > FIX #180
> 
> Hey, I was doing benchmarking on H100, of audio len around 18s. I found out that RTFx maxes out at a really low value around 120 at batch size 96. Can you share the benchmarking parameters or flags so that we can reproduce these results.

Hi, I am running the example `python examples/offline_inference/whisper.py` which measures the time to run ~2K audio clips:
```
Duration: 57.445557594299316
RPS: 35.65114668158841
```

**Comment by aryan-sarvam** [comment]
> Hi, I am running the example `python examples/offline_inference/whisper.py` which measures the time to run ~2K audio clips:

Hi, I ran the same exact script. The results were -

```Duration: 325.3493559360504
RPS: 6.294771950931872
```

Are there any other flags i have to use to reproduce the results?

**Comment by aryan-sarvam** [comment]
> Duration: 57.445557594299316
> RPS: 35.65114668158841

Hi @aurickq , I build the project from source and ran the same file. On h100, with exact same file, results are :-

```Duration: 309.14796566963196
RPS: 6.624659475160758
````

Can you tell if there are some other things to do? If possible Can you connect with me, sometime for proper setup. 

Thanks.

**Comment by mru4913** [comment]
@aurickq @DarkLight1337 is `docker run` ready for this model yet?

**Comment by DarkLight1337** [comment]
No, please see #12909

**Comment by mru4913** [comment]
@aryan-sarvam do you have any update ?
